Fix Parquet Parser missing column when reading parquet file

maytasm opened the pull request:
Fix Parquet Parser missing column when reading parquet file 

### Description

If you are ingesting Parquet using hadoop task, then you would set Parquet Hadoop Parser (https://druid.apache.org/docs/latest/ingestion/data-formats.html#parquet-hadoop-parser)
Looking at how things work, there is a DruidParquetReadSupport class that select the columns from the parquet schema that are used in the schema of the ingestion job. This is use to read in Parquet data for the map/reduce job.

DruidParquetReadSupport#getPartialReadSchema only consider the fields in the parseSpec->dimensionsSpec->dimensions, the ts field from parseSpec->timestampSpec->column, and the metrics’ requiredFields from dataSchema->metrics->fieldName
The issue: you have a transformSpec or a flattenSpec which creates a field that has a different name which does not exist in your parquet schema and you do not include the base field (that exist in parquet schema) in the dimensionsSpec. For example, your parquet file has column country, then you add a  transformSpec  that does something like "transforms": [ { "type": "expression", "name": "countryUpper", "expression": "upper(country)" }  . In the parseSpec->dimensionsSpec->dimensions, you only have countryUpper  (as you don’t want the original country field in your Druid datasource). The problem is that the DruidParquetReadSupport will not read in the country field (as country is not in the dimensionsSpec). As a result, you will not get  countryUpper  in your Druid datasource as Druid (map/reduce) job did not read in the country field.

This PR fix the above issue by determining the columns that the reader needs to read from parquet file by also looking at the flattenSpec and transformSpec.

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] added integration tests.
- [x] been tested in a test Druid cluster.

clintropolis commented on server/src/main/java/org/apache/druid/segment/indexing/ReaderUtils.java:
```diff
@@ -0,0 +1,143 @@
+    }
+
+    // Determine any fields we need to read from parquet file that is used in the dimensionsSpec
+    List<DimensionSchema> dimensionSchema = dimensionsSpec.getDimensions();
+    for (DimensionSchema dim : dimensionSchema) {
+      fieldsRequired.add(dim.getName());
+    }
+
+    // Determine any fields we need to read from parquet file that is used in the metricsSpec
+    for (AggregatorFactory agg : aggregators) {
+      fieldsRequired.addAll(agg.requiredFields());
+    }
```
since dimensions and aggregators can reference transforms, might want to exclude those here since the dim/aggregator would be referring to that instead of the source data.

I've also been thinking about about how to rework this, it feels to me like the object flattener should be fed a list of expected inputs computed from timestampSpec/transforms/dimensions/aggregators, which is sort of what you've got going on here, but I think the flattenSpec handling part should just be baked into flattener stuffs and handled when given this list of required inputs.

Note that I don't think this needs to be changed right now, this utility method is fine, just thinking ahead.

maytasm replied:
I have taken account that dimensions and aggregators can reference transforms by the line below,
`fieldsRequired.retainAll(fullInputSchema);`.

fullInputSchema is the columns of the input file. We are only retaining column names that exist in the input file which will remove all the transformed column.

maytasm replied:
Discussed offline and confirm that `fieldsRequired.retainAll(fullInputSchema);` at the end of the function should eliminate all fields that does not exist in the source input files and is sufficient.

clintropolis commented on extensions-core/parquet-extensions/example/compat/nested_array_struct.json:
```diff
@@ -45,7 +45,8 @@
             "dimensions": [
-              "i32_dec"
+              "i32_dec",
+              "primitive"
```
what are these changes for?

maytasm replied:
Some of our existing tests (such as CompatParquetInputTest#testReadNestedArrayStruct) read parquet using this configs. Reading is done via the RecordReader which used to read more than needed before this PR. For example, in CompatParquetInputTest#testReadNestedArrayStruct we expect the InputRow we read to have `primitive` column. However, the spec (before this change) does not need/require/consume the `primitive` column (`primitive` is not in the metric, dimension, transform, etc). To fix the test, we can either make it so that the spec requires the `primitive` (i.e. adding `primitive` to dimensionSpec) or remove the assert check on `primitive` column. Basically, our Reader is smarter and only read what is needed breaking some existing tests.

clintropolis reviewed (approved):
overall lgtm

samarthjain reviewed (commented):
LGTM. 

Can we add a test/tests where multiple files are ingested with one parquet file having a different than the other parquet file. The schema difference could be because of a column add/drop, column rename etc. You would likely need to tweak the TransformSpec in your test to handle renamed columns.

maytasm commented:
> LGTM.
> 
> Can we add a test/tests where multiple files are ingested with one parquet file having a different than the other parquet file. The schema difference could be because of a column add/drop, column rename etc. You would likely need to tweak the TransformSpec in your test to handle renamed columns.

@samarthjain Added an integration test with reading parquet files with multiple different schemas. 
```
File 1's columns: userRenamed, continent, country, added, deleted, deltaRenamed
File 2's columns: user, continentRenamed, country, added, delta
File 3's columns: user, continent, countryRenamed, deleted, delta
```
We have column renamed like column `user` is renamed to `userRenamed` in File 1. We also have columns dropped/added, where column `added` is dropped in File 3 and column `deleted` is dropped in File 2.

samarthjain commented on server/src/main/java/org/apache/druid/segment/indexing/ReaderUtils.java:
```diff
@@ -0,0 +1,143 @@
+    Set<String> fieldsRequired = new HashSet<>();
+
+    // We need to read timestamp column for Druid timestamp field
+    fieldsRequired.add(timestampSpec.getTimestampColumn());
```
Since flattenSpec is applied before timestampspec, do you think the order needs to be changed (I am not 100% sure, TBH)? Maybe worth adding a test where the timestamp column is not a top level field.

maytasm replied:
See ReaderUtilsTest#testGetColumnsRequiredForIngestionWithFlattenTimestamp

samarthjain commented:
+1, LGTM. Thanks, @maytasm!

maytasm commented:
Thanks for the review @samarthjain and @clintropolis
